axes/handlers/database.py

```diff
@@ -42,69 +35,30 @@ class AxesDatabaseHandler(AxesBaseHandler):  # pylint: disable=too-many-locals
     Signal handler implementation that records user login attempts to database and locks users out if necessary.
     """
 
-    def is_allowed(self, request: HttpRequest, credentials: Optional[Dict[str, Any]] = None) -> bool:
-        """
-        Check if the request or given credentials are already locked by Axes.
-
-        This function is called from
-
-        - function decorators defined in ``axes.decorators``,
-        - authentication backends defined in ``axes.backends``, and
-        - signal handlers defined in ``axes.handlers``.
-
-        This function checks the following facts for a given request:
+    def get_failures(self, request, credentials=None, attempt_time=None) -> int:
+        attempts = get_user_attempts(request, credentials, attempt_time)
+        return attempts.aggregate(Max('failures_since_start'))['failures_since_start__max'] or 0
 
-        1. Is the request IP address _blacklisted_? If it is, return ``False``.
-        2. Is the request IP address _whitelisted_? If it is, return ``True``.
-        4. Is the request HTTP method _whitelisted_? If it is, return ``True``.
-        3. Is the request user _whitelisted_? If it is, return ``True``.
-        5. Is failed authentication attempt always allowed to proceed? If it is, return ``True``.
-        6. Is failed authentication attempt count over the attempt limit? If it is, return ``False``.
-
-        Refer to the function source code for the exact implementation.
-        """
-
-        if is_client_ip_address_blacklisted(request):
+    def is_locked(self, request, credentials=None, attempt_time=None):
+        if is_user_attempt_whitelisted(request, credentials):
```

Comment:
Is this something that might be useful if it were moved to the parent function?

Reply:
I was thinking about the `is_user_attempt_whitelisted` for a while: the reason it is here is that it touches the database and checks the user model. 

If the cache handler should be purely cache based and not impose any load on the database on higher traffic sites, then having this in the cache backend as well (via the base class) produces redundant queries in all inheriting backends by default.

I was altogether considering removing this check as a whole since it seems to be a quite specialized check and should, at the very least, be more configurable than this. 

On the other hand, the handler based approach does offer a fairly customizable base for implementations, so maybe we are on good grounds here.